### ISSUE TITLE
ZipArchiveEntry.Name is wrong when FullName contains colon

### ISSUE THREAD

**Issue by theonlyremedy** [dotnet/corefx#41015]
I've created a ZipArchive using .Net Core, I'm facing issue when I try to unzip the file. The problem is caused by ZipArchiveEntry which has colon in its name. I agree that Windows file names can't contain colon, but its perfectly legal in other systems. When I iterate through the entries in the ZipArchive, the ZipArchiveEntry displays a wrong file name. Please refer attached screenshot, you can see that the FullName and Name show different file altogether.

![ZipArchiveEntryError](https://user-images.githubusercontent.com/11273401/64674306-21431600-d48e-11e9-9fd0-084358ecc6f8.png)

Please suggest a solution/fix for this.

**Comment by Wraith2** [comment]
The Name property will simply be the output of [GetFileName](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getfilename?view=netcore-3.0) which because you're calling it on windows regards the : as either illegal or part of the path root designation and returns the right most part after it.

**Comment by theonlyremedy** [comment]
But shouldn't it return the same no matter what the OS is? It should return the name of the file in the zip archive, not based on the OS or underlying filesystem.

**Comment by Wraith2** [comment]
Possibly, but it can't do that because the definition of path segments is different between the platforms and the zip file doesn't carry information about the origin rules.

**Comment by theonlyremedy** [comment]
Thank you @Wraith2 for your inputs. I have to use FullName and parse file name now. But Iterating through zip files should return same name across different systems, no matter I do it on Windows or Linux. I feel like am losing control of handling errors too. Let me go through ZipArchiveEntry's code to find a way if it can be solved. I would love some debate on this from other members too. Thanks again :)

**Comment by DaZombieKiller** [comment]
I definitely agree that the output of `Name` should not be platform-dependent.
The documentation for `ZipArchiveEntry.Name` states:

`The Name property contains the portion of the FullName property that follows the final directory separator character (\), and does not include the subdirectory hierarchy.`

This leaves no implication that it would be platform-specific or have special behaviour for colons.
If anything, this just seems like an oversight to me. A missing test case.

> Possibly, but it can't do that because the definition of path segments is different between the platforms and the zip file doesn't carry information about the origin rules.

This only strengthens the argument that this behaviour is wrong, because a zip file's definition of path segments is not necessarily the same as that of the platform it's being opened on.

**Comment by Wraith2** [comment]
I tend to agree and I've done come digging rather than guessing as I was before. It turns out that the zip file can contain knowledge of the source platform (not filesystem) which is in the [zip central directory](https://github.com/dotnet/corefx/blob/31a5bba0e79a5ae875f691d70ec3b9789691136d/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L56) and that it does later try to parse the FullName property in a platform specific manner
```csharp
https://github.com/dotnet/corefx/blob/31a5bba0e79a5ae875f691d70ec3b9789691136d/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L255
```

So if it's already trying to handle this as you want it to there's a problem somewhere because as far as I know it shouldn't care about the drive root designator in unix since they don't exist. Any chance you can provide a small sample zip file with the error so we can try to debug and see where it's going wrong?

**Comment by theonlyremedy** [comment]
Hey @Wraith2 & @DaZombieKiller Please have a look at the attached sample zip which was created using .Net Core code, when I checked the property in WinRAR, it showed HostOS as DOS, is this value added during creation and used during extraction? Attached screenshot below.
![image](https://user-images.githubusercontent.com/11273401/64887863-2cad6180-d687-11e9-8c01-e1a735d63535.png)

[archive.zip](https://github.com/dotnet/corefx/files/3611438/archive.zip)

**Comment by Wraith2** [comment]
Yup, the _versionMadeByPlatform is set to windows when I open that file. So perhaps when compressing it isn't being set. To check on the non-windows system where you're compressing the file can you try adding in this code and see what you get
```csharp
var zip = ZipFile.OpenRead(@"/path/to/archive.zip");
var archiveField = typeof(ZipArchiveEntry).GetField("_versionMadeByPlatform", BindingFlags.Instance | BindingFlags.NonPublic);
object value = archiveField.GetValue(zip.Entries[0]);
Console.WriteLine($"{value.GetType().Name}.{value:G}");
```

**Comment by theonlyremedy** [comment]
@Wraith2 Below is the code I used to create ZipArchive, I ran the code in both Windows 10 and Ubuntu 18.04 (using WSL).

```
class Program
{
  static void Main(string[] args)
  {
    string fileName = "archive.zip";
    CreateZip(fileName);
    System.Console.WriteLine("Done");
  }

  private static void CreateZip(string fileName)
  {
    using (FileStream fileStream = new FileStream(fileName, FileMode.Create))
    {
      using (ZipArchive zipArchive = new ZipArchive(fileStream, ZipArchiveMode.Create))
      {
        ZipArchiveEntry testFile = zipArchive.CreateEntry("testfolder/testfile: 2019-11-20.txt");
        using (StreamWriter streamWriter = new StreamWriter(testFile.Open()))
        {
          streamWriter.WriteLine("test archive entry with colon!");
        }
      }
    }
  }
}
```
The output archive's properties from WinRAR
![image](https://user-images.githubusercontent.com/11273401/64904447-da505d00-d6e7-11e9-8a59-7fd8da5dc632.png)

Archivers read this value, but I doubt they use it. When I open the archives created in Linux and Windows they both display same contents and extract properly, in Windows it replaces invalid characters with an underscore ('_') only while extracting, but they display names properly. I wish we had the same behavior in .Net too.

Code I used to read the files:
```
class Program
{
  static void Main(string[] args)
  {
    ReadZip(@"H:\archive_unix.zip");
    ReadZip(@"H:\archive_win.zip");
    System.Console.WriteLine("Done");
  }

  private static void ReadZip(string fileName)
  {
    using (FileStream fileStream = new FileStream(fileName, FileMode.Open))
    {
      using (ZipArchive zipArchive = new ZipArchive(fileStream, ZipArchiveMode.Read))
      {
        System.Console.WriteLine($"ZipFile: {fileName}");
        foreach(ZipArchiveEntry zipEntry in zipArchive.Entries)
        {
          FieldInfo madeByPlatformField = typeof(ZipArchiveEntry).GetField("_versionMadeByPlatform", BindingFlags.Instance | BindingFlags.NonPublic);
          object platfrom = madeByPlatformField.GetValue(zipEntry);
          System.Console.WriteLine($"{platfrom}\t{zipEntry.FullName}\t{zipEntry.Name}");
        }
      }
    }
  }
}
```
Output:
![image](https://user-images.githubusercontent.com/11273401/64904518-da049180-d6e8-11e9-916c-4218c96c2367.png)

I've also attached those ZipFiles for reference.
[archive_win.zip](https://github.com/dotnet/corefx/files/3612393/archive_win.zip)
[archive_unix.zip](https://github.com/dotnet/corefx/files/3612394/archive_unix.zip)

**Comment by carlossanlop** [comment]
@theonlyremedy @DaZombieKiller @Wraith2 Colon is a reserved character for filenames in Windows (documented [here](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file?redirectedfrom=MSDN#naming-conventions)).

Colon is a reserved character because it is used as the delimiter for Alternate Data Streams (ADS) in NTFS. You can read about them [here](https://blogs.technet.microsoft.com/askcore/2013/03/24/alternate-data-streams-in-ntfs/) and [here](http://www.ntfs.com/ntfs-multiple.htm).

Because of these OS specific restrictions, we don't want .NET Core to rename filenames without the developer's explicit request or consent. The developer should take care of these cases whenever they encounter a problem.

I created a zip file with .NET Core, containing a single text file named "beforecolon:aftercolon.txt". No problems arised when doing this. I was able to extract the file, but the suffix after the colon (including the colon) disappears from the extracted file's name, so the file shows up in File Explorer as "beforecolon".

I opened this file I created with different compression apps, to see what they do:

|Compression app|Result of drag/dropping file "beforecolon:aftercolon.txt" to Desktop|
|-|-|
|7-Zip|Renames it to "beforecolon_aftercolon.txt"|
|WinRAR|Renames it to "beforecolon_aftercolon.txt", then crashes.|
|Windows Explorer|Shows error dialog with message: "An unexpected error is keeping you from copying the file. If you continue to receive this error, you can use the error code to search the help with this problem. Error 0x80070057: The parameter is incorrect.".|

I also tested a couple of other languages:
|Language|Opening file and extracting "beforecolon:aftercolon.txt" without renaming|
|-|-|
|Python|Fails to extract file.|
|Java|Just like C#, extracts file as "beforecolon".|

For all 3 languages (C#, Java, Python), the workaround is the same: manually rename the file before extracting it.

Here's a code snippet with the workaround:

```cs
class Program
{
    static void Main(string[] args)
    {
        string directoryPath = @"C:\Users\yourusername\Desktop";
        string zipFileName = "file.zip";
        string zipFilePath = Path.Combine(directoryPath, zipFileName);

        // Create the zip file with a txt file containing a colon in the filename
        using (FileStream fs = new FileStream(zipFilePath, FileMode.Create))
        {
            using (ZipArchive za = new ZipArchive(fs, ZipArchiveMode.Create))
            {
                ZipArchiveEntry entry = za.CreateEntry("beforecolon:aftercolon.txt");
                using StreamWriter writer = new StreamWriter(entry.Open());
                writer.Write("Hello world");

                Console.WriteLine(entry.FullName);
            }
        }

        // Open it, extract the file replacing : with _
        using (FileStream fs = new FileStream(zipFilePath, FileMode.Open))
        {
            using (ZipArchive za = new ZipArchive(fs, ZipArchiveMode.Read))
            {
                foreach (var entry in za.Entries)
                {
                    Console.WriteLine(entry.FullName);

                    string extractFilePath = Path.Combine(directoryPath, entry.FullName.Replace(":", "_"));
                    
                    entry.ExtractToFile(extractFilePath, overwrite: true);
                    
                }
            }
        }
    }
}
```

CC @ericstj @JeremyKuhne in case they have any additional input.

**Comment by JeremyKuhne** [comment]
@carlossanlop did any of them actually create the alternate data stream of `aftercolon.txt` on `beforecolon`?

**Comment by carlossanlop** [comment]
@JeremyKuhne I didn't verify that, but I can check. What's the expectation?

**Comment by carlossanlop** [comment]
Talked to Jeremy offline. The expectation is that because of that colon, an ADS is created automatically.

I modified the code I shared above to **avoid** replacing `:` with `_`:

```cs
...
string extractFilePath = Path.Combine(directoryPath, entry.FullName);                 
...
```
This will create a file called "beforecolon", with empty contents, and with an attached ADS called "aftercolon.txt" with the text "Hello world" contained inside.

Then in Powershell 7 (which uses .NET Core), I can print the contents of both the main file (empty contents) and the alternate data stream ("Hello world") like this:

```ps
PS C:\> dir C:\Users\yourusername\desktop

    Directory: C:\Users\yourusername\Desktop

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---          2019-10-10  4:32 PM              0 beforecolon

PS C:\> type -LiteralPath \\?\C:\Users\yourusername\Desktop\beforecolon
PS C:\> type -LiteralPath \\?\C:\Users\yourusername\Desktop\beforecolon:aftercolon.txt
Hello world
```
